File: hash/js/rosetta-image-hash/src/internal/pilGaussianBlur.ts
/**
 * PIL GaussianBlur(radius=2) — byte-exact port of Pillow 10.4.0 BoxBlur.c.
 *
 * Pillow approximates a Gaussian by repeated box filtering. For radius=2 and
 * n=3 passes, the effective float box-radius is ~1.375 (from _gaussian_blur_radius).
 *
 * Algorithm:
 *   1. Compute floatRadius = _gaussian_blur_radius(sigma, passes).
 *   2. Compute integer weights ww/fw from floatRadius.
 *   3. Apply ALL n horizontal passes on the image rows.
 *   4. Transpose, apply ALL n horizontal passes (= vertical passes), transpose back.
 *
 * Boundary: edge replication (clamp). Intermediates stay in float/int64 range;
 * each 8-bit output pixel is (bulk + 2^23) >> 24, clamped to [0,255].
 *
 * Reference: spec/gaussian_blur_cases.json  (6 cases, all must pass byte-exact).
 */

/** _gaussian_blur_radius from BoxBlur.c */
function gaussianBlurRadius(sigma: number, passes: number): number {
  const sigma2 = (sigma * sigma) / passes;
  const L = Math.sqrt(12.0 * sigma2 + 1.0);
  const l = Math.floor((L - 1.0) / 2.0);
  const a =
    (2 * l + 1) * (l * (l + 1) - 3 * sigma2) /
    (6 * (sigma2 - (l + 1) * (l + 1)));
  return l + a;
}

/**
 * ImagingLineBoxBlur8 — one 1-D box-blur pass on a uint8 line.
 * lineIn must be Uint8Array or number[]; lineOut is a plain number[] of uint8.
 */
function boxBlurLine8(
  lineIn: Uint8Array | number[],
  n: number,
  radius: number,
  ww: number,
  fw: number,
  edgeA: number,
  edgeB: number,
): number[] {
  const lastx = n - 1;
  const lineOut: number[] = new Array(n);

  // Initial accumulator: from -(radius+1) to radius-1 with clamping
  let acc =
    lineIn[0] * (radius + 1);
  for (let x = 0; x < edgeA - 1; x++) {
    acc += lineIn[x];
  }
  acc += lineIn[lastx] * (radius - edgeA + 1);

  if (edgeA <= edgeB) {
    for (let x = 0; x < edgeA; x++) {
      acc += lineIn[x + radius] - lineIn[0];
      const bulk = acc * ww + (lineIn[0] + lineIn[x + radius + 1]) * fw;
      // Use >>> 24 (unsigned) to match C's UINT32 >> 24 — bulk can exceed 2^31.
      lineOut[x] = (bulk + (1 << 23)) >>> 24;
    }
    for (let x = edgeA; x < edgeB; x++) {
      acc += lineIn[x + radius] - lineIn[x - radius - 1];
      const bulk =
        acc * ww + (lineIn[x - radius - 1] + lineIn[x + radius + 1]) * fw;
      lineOut[x] = (bulk + (1 << 23)) >>> 24;
    }
    for (let x = edgeB; x <= lastx; x++) {
      acc += lineIn[lastx] - lineIn[x - radius - 1];
      const bulk =
        acc * ww + (lineIn[x - radius - 1] + lineIn[lastx]) * fw;
      lineOut[x] = (bulk + (1 << 23)) >>> 24;
    }
  } else {
    for (let x = 0; x < edgeB; x++) {
      acc += lineIn[x + radius] - lineIn[0];
      const bulk = acc * ww + (lineIn[0] + lineIn[x + radius + 1]) * fw;
      lineOut[x] = (bulk + (1 << 23)) >>> 24;
    }
    for (let x = edgeB; x < edgeA; x++) {
      acc += lineIn[lastx] - lineIn[0];
      const bulk = acc * ww + (lineIn[0] + lineIn[lastx]) * fw;
      lineOut[x] = (bulk + (1 << 23)) >>> 24;
    }
    for (let x = edgeA; x <= lastx; x++) {
      acc += lineIn[lastx] - lineIn[x - radius - 1];
      const bulk =
        acc * ww + (lineIn[x - radius - 1] + lineIn[lastx]) * fw;
      lineOut[x] = (bulk + (1 << 23)) >>> 24;
    }
  }

  return lineOut;
}

/**
 * Apply PIL GaussianBlur(radius=2) to a flat uint8 grayscale array (row-major, H×W).
 * Returns a new Uint8Array of the same size.
 *
 * @param src    Flat grayscale uint8 pixels, length = width * height.
 * @param width  Image width in pixels.
 * @param height Image height in pixels.
 */
export function pilGaussianBlur(
  src: Uint8Array,
  width: number,
  height: number,
): Uint8Array {
  const SIGMA = 2.0;
  const PASSES = 3;

  const floatRadius = gaussianBlurRadius(SIGMA, PASSES);
  const radius = Math.trunc(floatRadius); // (int) cast in C
  const ww = Math.trunc((1 << 24) / (floatRadius * 2 + 1)); // UINT32 division
  const fw = Math.trunc(((1 << 24) - (radius * 2 + 1) * ww) / 2);

  const edgeA_h = Math.min(radius + 1, width);
  const edgeB_h = Math.max(width - radius - 1, 0);
  const edgeA_v = Math.min(radius + 1, height);
  const edgeB_v = Math.max(height - radius - 1, 0);

  // Working buffer as number[] for arithmetic safety (avoid uint8 overflow mid-pass)
  let arr: number[] = Array.from(src);

  // All n horizontal passes
  for (let p = 0; p < PASSES; p++) {
    const tmp: number[] = new Array(width * height);
    for (let y = 0; y < height; y++) {
      const row = arr.slice(y * width, y * width + width);
      const out = boxBlurLine8(row, width, radius, ww, fw, edgeA_h, edgeB_h);
      for (let x = 0; x < width; x++) {
        tmp[y * width + x] = out[x];
      }
    }
    arr = tmp;
  }

  // All n vertical passes (Pillow transposes, blurs rows, transposes back)
  // Transposed array: shape (width, height) — transposed[x][y] = arr[y][x]
  let transposed: number[] = new Array(width * height);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      transposed[x * height + y] = arr[y * width + x];
    }
  }

  for (let p = 0; p < PASSES; p++) {
    const tmp: number[] = new Array(width * height);
    for (let x = 0; x < width; x++) {
      const col = transposed.slice(x * height, x * height + height);
      const out = boxBlurLine8(col, height, radius, ww, fw, edgeA_v, edgeB_v);
      for (let y = 0; y < height; y++) {
        tmp[x * height + y] = out[y];
      }
    }
    transposed = tmp;
  }

  // Transpose back
  const result = new Uint8Array(width * height);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      result[y * width + x] = transposed[x * height + y];
    }
  }

  return result;
}

File: hash/rust/rosetta-squint-hash/src/internal/pil_gaussian_blur.rs
//! Pillow-compatible GaussianBlur(radius=2) on 2-D grayscale uint8 images.
//!
//! Pillow implements gaussian blur as 3 passes of a fractional box blur
//! (ImagingGaussianBlur in BoxBlur.c), where the box radius is computed via:
//!
//!   sigma2 = radius^2 / passes   (passes = 3)
//!   L = sqrt(12 * sigma2 + 1)
//!   l = floor((L - 1) / 2)
//!   a = (2l+1)(l(l+1) - 3*sigma2) / (6*(sigma2 - (l+1)^2))
//!   float_radius = l + a
//!
//! Each pass is ImagingBoxBlurInterp: a box blur with fractional-pixel edges,
//! implemented with integer fixed-point arithmetic (24-bit shift):
//!   radius = floor(float_radius)
//!   ww = (1<<24) / (float_radius * 2 + 1)    (integer division of float)
//!   fw = ((1<<24) - (radius * 2 + 1) * ww) / 2
//!   output[x] = ((acc * ww + (left_edge + right_edge) * fw) + (1<<23)) >> 24
//!
//! Boundary condition: edge replication (clamp).
//! Passes: 3 horizontal then 3 vertical (via transpose trick).

/// Compute the box blur float-radius used by Pillow for gaussian approximation.
fn gaussian_box_radius(radius: f64, passes: f64) -> f64 {
    let sigma2 = radius * radius / passes;
    let l_raw = (12.0f64 * sigma2 + 1.0).sqrt();
    let l = ((l_raw - 1.0) / 2.0).floor();
    // Fractional part (from Gwosdek 2011 formula)
    let a_num = (2.0 * l + 1.0) * (l * (l + 1.0) - 3.0 * sigma2);
    let a_den = 6.0 * (sigma2 - (l + 1.0) * (l + 1.0));
    let a = a_num / a_den;
    l + a
}

/// One horizontal pass of the fixed-point fractional box blur (8-bit mode).
/// Matches Pillow's ImagingLineBoxBlur8 exactly.
fn line_box_blur8(line_in: &[u8], radius: usize, edge_a: usize, edge_b: usize, ww: u32, fw: u32) -> Vec<u8> {
    let lastx = line_in.len() - 1;
    let n = lastx + 1;
    let mut line_out = vec![0u8; n];

    // Initialize accumulator for virtual position x = -1.
    // acc = lineIn[0] * (radius + 1) + sum(lineIn[0..edgeA-1]) + lineIn[lastx] * max(0, radius - edgeA + 1)
    let mut acc: u32 = line_in[0] as u32 * (radius as u32 + 1);
    for x in 0..(edge_a.saturating_sub(1)) {
        acc += line_in[x] as u32;
    }
    // radius - edgeA + 1 can be 0 for large images
    if radius + 1 > edge_a {
        acc += line_in[lastx] as u32 * (radius - edge_a + 1) as u32;
    }

    if edge_a <= edge_b {
        // Left edge region: x in [0, edgeA)
        for x in 0..edge_a {
            acc = acc + line_in[x + radius] as u32 - line_in[0] as u32;
            let bulk = acc as u64 * ww as u64
                + (line_in[0] as u64 + line_in[x + radius + 1] as u64) * fw as u64;
            line_out[x] = ((bulk + (1 << 23)) >> 24) as u8;
        }
        // Middle region: x in [edgeA, edgeB)
        for x in edge_a..edge_b {
            acc = acc + line_in[x + radius] as u32 - line_in[x - radius - 1] as u32;
            let bulk = acc as u64 * ww as u64
                + (line_in[x - radius - 1] as u64 + line_in[x + radius + 1] as u64) * fw as u64;
            line_out[x] = ((bulk + (1 << 23)) >> 24) as u8;
        }
        // Right edge region: x in [edgeB, n)
        for x in edge_b..n {
            acc = acc + line_in[lastx] as u32 - line_in[x - radius - 1] as u32;
            let bulk = acc as u64 * ww as u64
                + (line_in[x - radius - 1] as u64 + line_in[lastx] as u64) * fw as u64;
            line_out[x] = ((bulk + (1 << 23)) >> 24) as u8;
        }
    } else {
        // Small image (edgeB < edgeA)
        for x in 0..edge_b {
            acc = acc + line_in[x + radius] as u32 - line_in[0] as u32;
            let bulk = acc as u64 * ww as u64
                + (line_in[0] as u64 + line_in[x + radius + 1] as u64) * fw as u64;
            line_out[x] = ((bulk + (1 << 23)) >> 24) as u8;
        }
        for x in edge_b..edge_a {
            acc = acc + line_in[lastx] as u32 - line_in[0] as u32;
            let bulk = acc as u64 * ww as u64
                + (line_in[0] as u64 + line_in[lastx] as u64) * fw as u64;
            line_out[x] = ((bulk + (1 << 23)) >> 24) as u8;
        }
        for x in edge_a..n {
            acc = acc + line_in[lastx] as u32 - line_in[x - radius - 1] as u32;
            let bulk = acc as u64 * ww as u64
                + (line_in[x - radius - 1] as u64 + line_in[lastx] as u64) * fw as u64;
            line_out[x] = ((bulk + (1 << 23)) >> 24) as u8;
        }
    }

    line_out
}

/// Apply one full horizontal pass over a 2-D image.
fn horizontal_box_blur(image: &[Vec<u8>], float_r: f64) -> Vec<Vec<u8>> {
    if image.is_empty() {
        return vec![];
    }
    let m = image[0].len();
    let radius = float_r as usize; // floor
    // Fixed-point weights
    let ww = ((1u32 << 24) as f64 / (float_r * 2.0 + 1.0)) as u32;
    let fw = ((1u32 << 24) - (radius * 2 + 1) as u32 * ww) / 2;
    let edge_a = (radius + 1).min(m);
    let edge_b = if m > radius + 1 { m - radius - 1 } else { 0 };

    image.iter().map(|row| line_box_blur8(row, radius, edge_a, edge_b, ww, fw)).collect()
}

/// Apply PIL GaussianBlur(radius=blur_radius) to a 2-D grayscale image.
///
/// Input:  `&[Vec<u8>]` — rows of uint8 pixels.
/// Output: `Vec<Vec<u8>>` — same shape, after 3 horizontal + 3 vertical passes.
pub fn apply(image: &[Vec<u8>], blur_radius: f64) -> Vec<Vec<u8>> {
    if image.is_empty() || image[0].is_empty() {
        return image.to_vec();
    }
    let float_r = gaussian_box_radius(blur_radius, 3.0);

    // 3 horizontal passes
    let mut result = image.to_vec();
    for _ in 0..3 {
        result = horizontal_box_blur(&result, float_r);
    }

    // 3 vertical passes via transpose trick
    let h = result.len();
    let w = result[0].len();
    // Transpose: shape (h, w) -> (w, h)
    let mut transposed: Vec<Vec<u8>> = (0..w).map(|x| (0..h).map(|y| result[y][x]).collect()).collect();
    for _ in 0..3 {
        transposed = horizontal_box_blur(&transposed, float_r);
    }
    // Transpose back: shape (w, h) -> (h, w)
    (0..h).map(|y| (0..w).map(|x| transposed[x][y]).collect()).collect()
}
